src/mlpack/methods/ann/layer/gru.hpp

```diff
@@ -0,0 +1,195 @@
+  GRUType();
+
+  /**
+   * Create the GRU layer object using the specified parameters.
+   *
+   * @param outSize The number of output units.
+   */
+  GRUType(const size_t outSize);
+
+  // Clone the GRUType object. This handles polymorphism correctly.
+  GRUType* Clone() const { return new GRUType(*this); }
+
+  // Copy the given GRUType object.
+  GRUType(const GRUType& other);
+  // Take ownership of the given GRUType object's data.
+  GRUType(GRUType&& other);
+  // Copy the given GRUType object.
+  GRUType& operator=(const GRUType& other);
+  // Take ownership of the given GRUType object's data.
+  GRUType& operator=(GRUType&& other);
+
+  virtual ~GRUType() { }
+
+  /**
+   * Reset the layer parameter. The method is called to
+   * assign the allocated memory to the internal learnable parameters.
+   */
+  void SetWeights(const MatType& weightsIn);
+
+  /**
+   * Ordinary feed forward pass of a neural network, evaluating the function
+   * f(x) by propagating the activity forward through f.
+   *
+   * @param input Input data used for evaluating the specified function.
+   * @param output Resulting output activation.
+   */
+  void Forward(const MatType& input, MatType& output);
+
+  /**
+   * Ordinary feed backward pass of a neural network, calculating the function
+   * f(x) by propagating x backwards trough f. Using the results from the feed
+   * forward pass.
+   *
+   * @param input The input data (x) given to the forward pass.
+   * @param output The propagated data (f(x)) resulting from Forward()
+   * @param gy Propagated error from next layer.
+   * @param g Matrix to store propagated error in for previous layer.
+   */
+  void Backward(const MatType& /* input */,
+                const MatType& output,
+                const MatType& gy,
+                MatType& g);
+
+  /*
+   * Calculate the gradient using the output delta and the input activation.
+   *
+   * @param input Original input data provided to Forward().
+   * @param error Error as computed by `Backward()`.
+   * @param gradient Matrix to store the gradients in.
+   */
+  void Gradient(const MatType& input,
+                const MatType& /* error */,
+                MatType& gradient);
+
+  // Get the parameters.
+  MatType const& Parameters() const { return weights; }
+  // Modify the parameters.
+  MatType& Parameters() { return weights; }
+
+  // Get the total number of trainable parameters.
+  size_t WeightSize() const;
+
+  // Get the total number of recurrent state parameters.
+  size_t RecurrentSize() const;
+
+  // Given a properly set InputDimensions(), compute the output dimensions.
+  void ComputeOutputDimensions()
+  {
+    inSize = this->inputDimensions[0];
+    for (size_t i = 1; i < this->inputDimensions.size(); ++i)
+      inSize *= this->inputDimensions[i];
+    this->outputDimensions = std::vector<size_t>(this->inputDimensions.size(),
+        1);
+
+    // The GRU layer flattens its input.
+    this->outputDimensions[0] = outSize;
```

Comment:
```suggestion
    // The GRU layer flattens its input.
    this->outputDimensions = std::vector<size_t>({ outSize });
```
Maybe a slight simplification?

Reply:
Doing it this way preserves the number of dimensions. I'm not completely certain if there is a reason for doing this but it is how every other layer does it.

Reply:
You're right---I thought that the result there was identical, but it's actually not... which is probably why the other layers were written like that too.  I think it may have been too late in the day when I reviewed this and I wasn't thinking clearly... :)